### ISSUE TITLE
Linker error

### ISSUE THREAD

**Issue by o3o** [rejectedsoftware/vibe.d#1862]
Hi,
after upgrade:
```
$ cd vibe.d
$ git pull
$ git describe
v0.8.1-alpha.2-9-gfc2c3c00

$ dub upgrade
```
I've a lot of linker errors:
```
$ cd examples/app_skeleton/
$ dub upgrade
$ dub build


app-skeleton ~master: building configuration "application"...
Compiling Diet HTML template error.dt...
Compiling Diet HTML template about.dt...
Compiling Diet HTML template home.dt...
Linking...
/usr/local/bin/ld: error: /usr/lib/libphobos2.a(object_1_257.o): requires dynamic R_X86_64_32 reloc against '_D9Exception7__ClassZ' which may overflow at runtime; recompile with -fPIC
/usr/local/bin/ld: error: /usr/lib/libphobos2.a(object_7_32e.o): requires dynamic R_X86_64_32 reloc against '_D8TypeInfo7__ClassZ' which may overflow at runtime; recompile with -fPIC
/usr/local/bin/ld: error: /usr/lib/libphobos2.a(object_7_32e.o): requires dynamic R_X86_64_32 reloc against '_D8TypeInfo7__ClassZ' which may overflow at runtime; recompile with -fPIC
/usr/local/bin/ld: error: /usr/lib/libphobos2.a(object_a_66e.o): requires dynamic R_X86_64_32 reloc which may overflow at runtime; recompile with -fPIC
/usr/local/bin/ld: error: /usr/lib/libphobos2.a(object_a_66e.o): requires dynamic R_X86_64_32 reloc against '_D16TypeInfo_Pointer7__ClassZ' which may overflow at runtime; recompile with -fPIC
....
```
My environment:
```
$ uname --all
Linux abe 4.11.9-1-ARCH #1 SMP PREEMPT Wed Jul 5 18:23:08 CEST 2017 x86_64 GNU/Linux

$ dmd --version
DMD64 D Compiler v2.074.0
Copyright (c) 1999-2017 by Digital Mars written by Walter Bright

$ dub --version
DUB version 1.3.0, built on May 23 2017
```
I use  [gold linker](https://en.wikipedia.org/wiki/Gold_&#40;linker&#41;)
```
$ ls -l /usr/local/bin/ld 
lrwxrwxrwx 1 root root 16 Apr  4 11:15 /usr/local/bin/ld -> /usr/bin/ld.gold

$ ld -v
GNU gold (GNU Binutils 2.28.0.20170506) 1.14
```

Thank you

**Comment by Geod24** [comment]
Another PIC issue. Can you compile a simple hello world program ? What is the content of your dmd config file ? If you are compiling with `-fPIC`, you should have `-defaultlib=libphobos2.so` as well.

**Comment by wilzbach** [comment]
I would guess that you are using Arch Linux?
See http://forum.dlang.org/post/vjpawnhzfvxffvsiojuf@forum.dlang.org or http://forum.dlang.org/post/vjpawnhzfvxffvsiojuf@forum.dlang.org

**Comment by o3o** [comment]
@Geod24 
Thank you Mathias..

> Can you compile a simple hello world program ?

No I can't compile, ...so isn't a vibe.d issue...

> What is the content of your dmd config file ?

```
$ cat /etc/dmd.conf 
[Environment32]
DFLAGS=-I/usr/include/dlang/dmd -L-L/usr/lib32 -L--export-dynamic -fPIC
[Environment64]
DFLAGS=-I/usr/include/dlang/dmd -L-L/usr/lib -L--export-dynamic -fPIC
```

If I add `-defaultlib=libphobos2.so`:
```
[Environment64]
DFLAGS=-I/usr/include/dlang/dmd -L-L/usr/lib -L--export-dynamic -fPIC -defaultlib=libphobos2.so
```

it works!

**Comment by o3o** [comment]
@wilzbach Thank you Sebastian
> I would guess that you are using Arch Linux?
Yes I've Archlinux. 

So, solution proposed by @Geod24 works, ...but which is the best solution?

**Comment by wilzbach** [comment]
> So, solution proposed by @Geod24 works, ...but which is the best solution?

Please see http://forum.dlang.org/post/vjpawnhzfvxffvsiojuf@forum.dlang.org

tl;dr:
- The Arch TU maintainers have finally rebuilt Phobos with `PIE=1` which was showing up in your error message -> simply "pacman -Syu" and everything should work
- Regarding "best solution": depends on your use-case. I prefer static linking of Phobos because I can rsync my binaries to other machines :)

**Comment by o3o** [comment]
> The Arch TU maintainers have finally rebuilt Phobos with PIE=1 which was showing up in your error message -> simply "pacman -Syu" and everything should work

Great!

>   Regarding "best solution": depends on your use-case. I prefer static linking of Phobos because I can rsync my binaries to other machines :)
Of course "best" is an arguable word.

But I do not understand your preference, could you please describe me:
- *static linking* means without `-fPIC?
- why with static linking you can rsync binary? 

Thank you very much.!

**Comment by ZombineDev** [comment]
@o3o `-fPIC` is orthogonal to static linking, but necessary for dynamic linking. I think what @wilzbach meant was compiling everything with `-fPIC` and then static linking everything (druntime, phobos, vibe.d and his app) into one single binary, so he can just copy and paste it. When using dynamic linking `libphobos2.so` is a separate file located in a different place than statically linked vibe.d + his app which means he needs some distro/platform specific code to locate `libphobos2.so` when copying his app because in general you need to use the exact same version of `libphobos2.so` which you used to build your project. You can't build your project with one version of `libphobos2.so` and then deploy it on another computer with another version of `libphobos2.so`, because most D shared libraries have no backwards compatibility.

**Comment by o3o** [comment]
@ZombineDev
Thank you Petar, now it's more clear.
